src/harness/sas.py

```diff
@@ -332,3 +342,113 @@ def _GetDefaultAdminSSLCertPath(self):
+   }
+   schema of full_activity {full activity dump object}
+   schema of cbsd_activity {"id": {cbsd activity dump object}}
+   schema of PPA_activity {"id": {ppa activity dump object}}
+   schema of esc_activity {"id": {esc activity dump object}}
+
+  """
+  def __init__(self, server_details):
+    self._server_details = server_details
+    self.parameters = {}
+
+
+  def _handleRequest(self, getSetupParameters):
+    return lambda *args: HttpServerHandler(getSetupParameters, *args)
+
+  def _getSetupParameters(self):
+    return self.parameters
+
+
+  def setupFullActivity(self, full_activity_dump=None):
+    self.parameters['full_activity'] = full_activity_dump
+
+  def setupCbsdActivity(self, unique_id, Cbsd_activity_dump = None):
+    self.parameters['cbsd_activity'][unique_id] = Cbsd_activity_dump
+
+  def setupPPAActivity(self, unique_id, PPA_activity_dump = None):
+    self.parameters['PPA_activity'][unique_id] = PPA_activity_dump
+
+  def setupEscActivity(self,unique_id, esc_activity_dump = None):
+    self.parameter['esc_activity'][unique_id] = esc_activity_dump
+
+  def getBaseUrl(self):
+    return 'https://%s:%d' % (self._server_details['baseUrl'], self._server_details['port'])
+
+  def StartServer(self, ca_cert):
+    request_handler = self._handleRequest(self._getSetupParameters)
+    self.server = HTTPServer((self._server_details['baseUrl'], self._server_details['port']),request_handler)
+    cert_file = 'server.cert'
+    key_file = 'server.key'
+    ciphers = ['AES128-GCM-SHA256', 'AES256-GCM-SHA384', 'ECDHE-RSA-AES128-GCM-SHA256', 'ECDHE-ECDSA-AES256-GCM-SHA384', 'ECDHE-RSA-AES128-GCM-SHA256']
+    self.server.socket = ssl.wrap_socket(
+      self.server.socket,
+      certfile=cert_file,
+      keyfile=key_file,
+      ca_certs=ca_cert,
+      cert_reqs=ssl.CERT_REQUIRED,
+      ssl_version=ssl.PROTOCOL_TLSv1_2,
+      ciphers=':'.join(ciphers),
+      server_side=True
+      )
+    logging.info('Started Test Harness Server at %s' % self.getBaseUrl())
+    # Start Server
+    self.server.serve_forever()
+
+  def StopServer(self):
+    logging.info('Stopped Test Harness Server')
+    self.server.shutdown()
+
+class HttpServerHandler(BaseHTTPRequestHandler):
+  def __init__(self, getSetupParameters, *args):
+    self._parameters = getSetupParameters()
+    BaseHTTPRequestHandler.__init__(self, *args)
+
+  def do_GET(self):
+    """Handles Pull/GET Request and returns Path of the Request to callback Method"""
+    if "dump" in self.path:
+      self.send_response(200)
+      self.send_header('Content-type', 'application/json')
+      self.end_headers()
+      self.wfile.write(json.dumps(self._parameters['full_activity']))
+    elif "cbsd_activity" in self.path:
+      self.send_response(200)
+      self.send_header('Content-type', 'application/json')
+      self.end_headers()
+      id = self.path.split('/')[-1]
+      if id not in self._parameters['cbsd_activity']:
+        self.send_response(404)
```

Comment:
Does this need to be send_error() instead?

Also it's not clear to me what happens if you call send_respose() twice (or send_response() and then send_error()). It seems less error-prone to move the "if id not in..." check to the top of this elif block.